pkg/controller/util/upgrade.go

```diff
@@ -0,0 +1,76 @@
+package util
+
+import (
+	"fmt"
+	"github.com/Masterminds/semver"
+	"github.com/go-logr/logr"
+	"github.com/tidwall/gjson"
+	"io/ioutil"
+	"net/http"
+	"strings"
+)
+
+const (
+	ProxyImageName = "proxy"
+
+	CollectorImageName = "wavefront-kubernetes-collector"
+
+	ImagePrefix = "wavefronthq/"
+)
+
+func GetLatestVersion(crImageName string, currentVersion string, enableAutoUpgrade bool, reqLogger logr.Logger) (string, error) {
+	if !enableAutoUpgrade {
+		return currentVersion, nil
+	}
+
+	// "latest" effectively renders auto upgrade useless.
+	if currentVersion == "latest" {
+		return currentVersion, nil
+	}
+
+	// The last 20 tags should be good. Don't expect customers to be using a really old version of CR.
+	url := "https://registry.hub.docker.com/v2/repositories/wavefronthq/" + crImageName + "/tags/?page_size=20"
+	resp, err := http.Get(url)
+	if err != nil {
+		return "", err
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != http.StatusOK {
+		return "", fmt.Errorf("unexpected http GET status: %s", resp.Status)
+	}
+	contents, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return "", err
+	}
+
+	// The below will get us the versions from json.
+	// Ex for Proxy: [latest 5.5 5.1 4.38 4.36 4.35 4.34 4.33 4.32 4.31]
+	versions := gjson.Get(string(contents), "results.#.name")
```

Comment:
Break this into a separate function that processes the result of the call to the docker hub API and add unit tests around this by passing in contents with different image tags etc. For example if we tag as `v1.0.3` vs `1.0.3` and such.

Reply:
This code just unmarshall's the docker "tag" string. I don't think unit testing this is needed(gjson.get() does nothing different based on the value of the tag itself)

As a side note, The Masterminds/semver library we use to compare tags support only comparing semantically versioned tags. If we tag our docker images with non semantic versioned tags (Ex: version1.0.3) , our current auto upgrade comparison will not work. 
An exception is when Tag is prefixed "v" . The library does work optionally with any semantic versioned tag prefix with "v" - so currently comparing "v1.0.3" and "1.0.3" will both work.